lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -1,23 +1,126 @@
-"""TODO: Add a proper docstring here.
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
 
-This is a placeholder docstring for this charm library. Docstrings are
-presented on Charmhub and updated whenever you push a new version of the
-library.
+"""Library for the ip-router integration
 
-Complete documentation about creating and documenting libraries can be found
-in the SDK docs at https://juju.is/docs/sdk/libraries.
+This library contains the Requires and Provides classes for interactions through the 
+ip-router interface
 
-See `charmcraft publish-lib` and `charmcraft fetch-lib` for details of how to
-share and consume charm libraries. They serve to enhance collaboration
-between charmers. Use a charmer's libraries for classes that handle
-integration with their charm.
+## Getting Started
 
-Bear in mind that new revisions of the different major API versions (v0, v1,
-v2 etc) are maintained independently.  You can continue to update v0 and v1
-after you have pushed v3.
+From a charm directory, fetch the library using `charmcraft`:
 
-Markdown is supported, following the CommonMark specification.
-"""
+```shell
+charmcraft fetch-lib charms.ip_router_interface.v0.ip_router_interface
+```
+
+### Provider charm
+This example provider charm is all we need to listen to ip-router requirer requests.
+The ip-router provider fulfills the routing function for multiple charms that are
+requirers of the ip-router interface. For that reason, this charm will continuously
+track and update all of the connected requirers and the networks and routes they've 
+requested, which it does in a routing table. As new charms are connected and disconnected
+to the relation, this routing table is automatically adds and removes the dependent
+networks.
+
+The library handles the listening and synchronization for all of the ip-router network
+requests internally, which means as the charm author you don't need to worry about any
+of the business logic of validating or orchestrating the relation network.
+
+You can also listen to the `routing_table_updated` event for convenience.
+
+
+```python
+import logging, json
+import ops
+from charms.ip_router_interface.v0.ip_router_interface import *
+
+class SimpleIPRouteProviderCharm(ops.CharmBase):
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.RouterProvider = RouterProvides(charm=self)
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.routing_table_updated, self._routing_table_updated)
+        self.framework.observe(self.on.get_routing_table_action, self._action_get_routing_table)
+
+    def _on_install(self, event: ops.InstallEvent):
+        pass
```

Comment:
Why do we need this?

Reply:
My charms would throw a hook failed error without ops listening to the install, even if we do nothing with it.

Reply:
Changed it to set status as active. Done.